more_itertools/more.py

```diff
@@ -1634,35 +1634,148 @@ def numeric_range(*args):
+        if argc == 1:
+            self._stop, = args
+            self._start = type(self._stop)(0)
+            self._step = 1
+        elif argc == 2:
+            self._start, self._stop = args
+            self._step = 1
+        elif argc == 3:
+            self._start, self._stop, self._step = args
+        else:
+            if argc == 0:
+                err_msg = 'numeric_range expected 1 argument, got {}'
+            else:
+                err_msg = 'numeric_range expected at most 3 arguments, got {}'
+            raise TypeError(err_msg.format(argc))
+
+        self._zero = type(self._step)(0)
+        if self._step == self._zero:
+            raise ValueError('numeric_range() arg 3 must not be zero')
+        self._growing = self._step > self._zero
+
+    def __bool__(self):
+        if self._growing:
+            return self._start < self._stop
+        else:
+            return self._start > self._stop
+
+    def __contains__(self, elem):
+        if self._growing:
+            if self._start <= elem < self._stop:
+                return (elem - self._start) % self._step == self._zero
+        else:
+            if self._start >= elem > self._stop:
+                return (self._start - elem) % (-self._step) == self._zero
+
+        return False
+
+    def __eq__(self, other):
+        return (self._start == other._start
+                and self._step == other._step
+                and self._get_by_index(-1) == other._get_by_index(-1))
+
+    def __getitem__(self, key):
+        if isinstance(key, int):
+            return self._get_by_index(key)
+        elif isinstance(key, slice):
+            start = self._start if key.start is None \
+                else self._get_by_index(key.start)
+            stop = self._stop if key.stop is None \
+                else self._get_by_index(key.stop)
+            step = self._step if key.step is None else key.step * self._step
+            return numeric_range(start, stop, step)
+        else:
+            raise TypeError(
+                "numeric range indices must be integers or slices, not str")
+
+    def __hash__(self):
+        if self:
+            return hash((self._start, self._get_by_index(-1), self._step))
+        else:
+            return self._EMPTY_HASH
+
+    def __iter__(self):
+        elem = self._start + self._zero
+        if self._growing:
+            while elem < self._stop:
+                yield elem
+                elem += self._step
+        else:
+            while elem > self._stop:
+                yield elem
+                elem += self._step
+
+    def __len__(self):
+        if self._growing:
+            start = self._start
+            stop = self._stop
+            step = self._step
+        else:
+            start = self._stop
+            stop = self._start
+            step = -self._step
+
+        distance = stop - start
+        if distance <= 0:
+            return 0
+
+        r = distance % step
+        q = int(distance / step)
```

Comment:
Why not `divmod` here?

Reply:
Because I initially thought that `5.0 // 1.5` would return `3` and not `3.0`. The code was:

        r = distance % step
        q = distance // step

And then I added a `int` cast and removed the second slash. I'll use `divmod`. Note that there is a subtle difference:

```
>>> divmod(5.0, -1.5)
(-4.0, -1.0)
>>> (5.0//-1.5, 5.0%-1.5)
(-4.0, -1.0)
```

But:
```
>>> (int(5.0/-1.5), 5.0%-1.5)
(-3, -1.0)
```

We do not have this problem here because `distance` and `step` are both positive. I propose an `assert` to document this:

        assert distance > 0 and step > 0

Reply:
Thanks for the explanation. No need for the assert, but could add this as a comment.

Reply:
Done.